File: daemon-python/arcanos/openai/request_builders.py
"""
Standardized Request Builder Patterns for Python CLI Agent

Provides reusable request builders for all OpenAI API operations:
- Chat completions (with ARCANOS routing message)
- Vision requests
- Audio transcription
- Image generation
- Embeddings

Features:
- Railway-native patterns (stateless, deterministic)
- Consistent request structure
- Type-safe builders
- ARCANOS routing message injection
- Audit trail for all requests
"""

from typing import Any, Dict, List, Optional, Union

from ..config import Config


def _normalize_content_to_text(content: Any) -> str:
    """
    Purpose: Normalize mixed message content payloads into plain text.
    Inputs/Outputs: arbitrary message content -> normalized string.
    Edge cases: Unknown content parts resolve to empty strings.
    """
    if isinstance(content, str):
        return content

    if isinstance(content, list):
        normalized_parts: List[str] = []
        for part in content:
            if not isinstance(part, dict):
                continue
            #audit Assumption: text-like parts should preserve deterministic prompt content; risk: silent data loss in multimodal lists; invariant: text and input_text parts are extracted when present; handling: collect supported text fields only.
            if part.get("type") in {"text", "input_text"} and isinstance(part.get("text"), str):
                normalized_parts.append(part["text"])
        return "\n".join(part for part in normalized_parts if part)

    return ""


def extract_response_text(response: Any, fallback: str = "") -> str:
    """
    Purpose: Extract normalized text output from Responses API payloads.
    Inputs/Outputs: OpenAI response object -> output text string.
    Edge cases: Falls back to scanning output content when output_text shortcut is absent.
    """
    output_text = getattr(response, "output_text", None)
    if isinstance(output_text, str) and output_text.strip():
        return output_text.strip()

    output = getattr(response, "output", None)
    if not isinstance(output, list):
        return fallback

    for item in output:
        if not isinstance(item, dict):
            continue
        content_list = item.get("content", [])
        if not isinstance(content_list, list):
            continue
        for content_item in content_list:
            if not isinstance(content_item, dict):
                continue
            if content_item.get("type") == "output_text" and isinstance(content_item.get("text"), str):
                candidate = content_item["text"].strip()
                if candidate:
                    return candidate
    return fallback


def build_responses_request(
    prompt: str,
    system_prompt: Optional[str] = None,
    model: Optional[str] = None,
    max_tokens: Optional[int] = None,
    temperature: Optional[float] = None,
    top_p: Optional[float] = None,
    conversation_history: Optional[List[Dict[str, str]]] = None,
) -> Dict[str, Any]:
    """
    Purpose: Build canonical Responses API request payload for text interactions.
    Inputs/Outputs: prompt + generation controls -> responses.create payload.
    Edge cases: Missing history/system prompts degrade to single user message input.
    """
    resolved_temperature = Config.TEMPERATURE if temperature is None else temperature
    resolved_max_tokens = Config.MAX_TOKENS if max_tokens is None else max_tokens

    input_items: List[Dict[str, Any]] = []
    if conversation_history:
        for conversation_item in conversation_history[-5:]:
            #audit Assumption: conversation history keys follow user/ai schema; risk: malformed history breaks context reconstruction; invariant: only string values are emitted into response input; handling: guard each key before append.
            if isinstance(conversation_item.get("user"), str):
                input_items.append({
                    "role": "user",
                    "content": [{"type": "input_text", "text": conversation_item["user"]}]
                })
            if isinstance(conversation_item.get("ai"), str):
                input_items.append({
                    "role": "assistant",
                    "content": [{"type": "input_text", "text": conversation_item["ai"]}]
                })

    input_items.append({
        "role": "user",
        "content": [{"type": "input_text", "text": prompt}]
    })

    payload: Dict[str, Any] = {
        "model": model or Config.OPENAI_MODEL,
        "input": input_items,
        "temperature": resolved_temperature,
        "max_output_tokens": resolved_max_tokens,
    }

    if system_prompt:
        payload["instructions"] = system_prompt
    if top_p is not None:
        payload["top_p"] = top_p

    return payload


def build_vision_responses_request(
    prompt: str,
    image_base64: str,
    model: Optional[str] = None,
    max_tokens: Optional[int] = None,
    temperature: Optional[float] = None,
) -> Dict[str, Any]:
    """
    Purpose: Build canonical Responses API request payload for vision interactions.
    Inputs/Outputs: prompt + base64 image -> responses.create payload.
    Edge cases: Uses PNG data URL format when MIME metadata is unavailable.
    """
    resolved_temperature = Config.TEMPERATURE if temperature is None else temperature
    resolved_max_tokens = Config.MAX_TOKENS if max_tokens is None else max_tokens

    return {
        "model": model or Config.OPENAI_VISION_MODEL,
        "input": [
            {
                "role": "user",
                "content": [
                    {"type": "input_text", "text": prompt},
                    {"type": "input_image", "image_url": f"data:image/png;base64,{image_base64}"},
                ],
            }
        ],
        "temperature": resolved_temperature,
        "max_output_tokens": resolved_max_tokens,
    }
def build_chat_completion_request(
    prompt: str,
    system_prompt: Optional[str] = None,
    model: Optional[str] = None,
    max_tokens: Optional[int] = None,
    temperature: Optional[float] = None,
    top_p: Optional[float] = None,
    frequency_penalty: Optional[float] = None,
    presence_penalty: Optional[float] = None,
    conversation_history: Optional[List[Dict[str, str]]] = None,
    include_routing_message: bool = True
) -> Dict[str, Any]:
    """
    Builds a chat completion request with ARCANOS routing message
    
    Automatically prepends ARCANOS routing message to ensure proper
    model routing and behavior. This is the standard way to create
    chat completion requests in the codebase.
    """
    # Build messages array
    messages: List[Dict[str, str]] = []
    
    # Add system prompt if provided
    if system_prompt:
        messages.append({"role": "system", "content": system_prompt})
    
    # Add conversation history
    if conversation_history:
        for conv in conversation_history[-5:]:  # Last 5 conversations
            if "user" in conv:
                messages.append({"role": "user", "content": conv["user"]})
            if "ai" in conv:
                messages.append({"role": "assistant", "content": conv["ai"]})
    
    # Add user prompt
    messages.append({"role": "user", "content": prompt})
    
    # Build request payload
    # //audit assumption: callers may intentionally pass 0/0.0 generation values; risk: truthy fallback would override explicit intent; invariant: defaults only when values are omitted (None); strategy: explicit None checks before payload construction.
    resolved_temperature = Config.TEMPERATURE if temperature is None else temperature
    resolved_max_tokens = Config.MAX_TOKENS if max_tokens is None else max_tokens

    request_payload: Dict[str, Any] = {
        "model": model or Config.OPENAI_MODEL,
        "messages": messages,
        "temperature": resolved_temperature,
        "max_tokens": resolved_max_tokens,
    }
    
    if top_p is not None:
        request_payload["top_p"] = top_p
    if frequency_penalty is not None:
        request_payload["frequency_penalty"] = frequency_penalty
    if presence_penalty is not None:
        request_payload["presence_penalty"] = presence_penalty
    
    return request_payload


def build_vision_request(
    prompt: str,
    image_base64: str,
    model: Optional[str] = None,
    max_tokens: Optional[int] = None,
    temperature: Optional[float] = None,
    detail: str = "auto"
) -> Dict[str, Any]:
    """
    Builds a vision request for image analysis
    
    Creates a properly formatted vision request with image data
    and user prompt.
    """
    messages = [
        {
            "role": "user",
            "content": [
                {"type": "text", "text": prompt},
                {
                    "type": "image_url",
                    "image_url": {
                        "url": f"data:image/png;base64,{image_base64}",
                        "detail": detail
                    }
                }
            ]
        }
    ]
    
    # //audit assumption: explicit zero generation values can be intentional for deterministic behavior; risk: truthy fallback drops explicit value; invariant: defaults apply only for None; strategy: explicit None checks.
    resolved_temperature = Config.TEMPERATURE if temperature is None else temperature
    resolved_max_tokens = Config.MAX_TOKENS if max_tokens is None else max_tokens

    return {
        "model": model or Config.OPENAI_VISION_MODEL,
        "messages": messages,
        "temperature": resolved_temperature,
        "max_tokens": resolved_max_tokens,
    }


def build_transcription_request(
    audio_file: Any,  # File-like object or bytes
    filename: str,
    model: Optional[str] = None,
    language: Optional[str] = None,
    response_format: str = "json",
    temperature: Optional[float] = None
) -> Dict[str, Any]:
    """
    Builds a transcription request for audio processing
    
    Creates a properly formatted transcription request with audio file.
    """
    request_params: Dict[str, Any] = {
        "file": audio_file,
        "model": model or Config.OPENAI_TRANSCRIBE_MODEL,
        "response_format": response_format
    }
    
    if language:
        request_params["language"] = language
    
    if temperature is not None:
        request_params["temperature"] = temperature
    
    return request_params


def build_image_request(
    prompt: str,
    size: str = "1024x1024",
    model: Optional[str] = None,
    quality: str = "standard",
    n: int = 1,
    response_format: str = "b64_json"
) -> Dict[str, Any]:
    """
    Builds an image generation request
    
    Creates a properly formatted image generation request.
    """
    return {
        # Prefer configurable image model when available, fall back to
        # the default literal used in main.
        "model": model or Config.OPENAI_IMAGE_MODEL,
        "prompt": prompt,
        "size": size,
        "quality": quality,
        "n": n,
        "response_format": response_format
    }


def build_embedding_request(
    input_text: Union[str, List[str]],
    model: str = "text-embedding-3-small",
    user: Optional[str] = None
) -> Dict[str, Any]:
    """
    Builds an embedding request
    
    Creates a properly formatted embedding request.
    """
    request_params: Dict[str, Any] = {
        "model": model,
        "input": input_text
    }
    
    if user:
        request_params["user"] = user
    
    return request_params


__all__ = [
    "_normalize_content_to_text",
    "extract_response_text",
    "build_responses_request",
    "build_vision_responses_request",
    "build_chat_completion_request",
    "build_vision_request",
    "build_transcription_request",
    "build_image_request",
    "build_embedding_request"
]

File: src/services/openai/requestBuilders.ts
/**
 * Standardized Request Builder Patterns
 * 
 * Provides reusable request builders for all OpenAI API operations:
 * - Chat completions (with ARCANOS routing message)
 * - Vision requests
 * - Audio transcription
 * - Image generation
 * - Embeddings
 * 
 * Features:
 * - Railway-native patterns (stateless, deterministic)
 * - Consistent request structure
 * - Type-safe builders
 * - ARCANOS routing message injection
 * - Audit trail for all requests
 * 
 * @module requestBuilders
 */

import type OpenAI from 'openai';
import type {
  Response as OpenAIResponse,
  ResponseCreateParamsNonStreaming,
  ResponseInput
} from 'openai/resources/responses/responses';
import type { ChatCompletionMessageParam, ChatCompletionResponseFormat, ImageSize } from './types.js';
import { ARCANOS_ROUTING_MESSAGE } from './unifiedClient.js';
import { getTokenParameter } from "@shared/tokenParameterHelper.js";
import { buildSystemPromptMessages } from "@shared/messageBuilderUtils.js";
import { DEFAULT_IMAGE_SIZE, IMAGE_GENERATION_MODEL, ROUTING_MAX_TOKENS } from './config.js';
import { OPENAI_COMPLETION_DEFAULTS } from "./constants.js";

/**
 * Chat completion request parameters
 */
export interface ChatParams {
  /** User prompt/message */
  prompt: string;
  /** System prompt (optional, ARCANOS routing message will be prepended) */
  systemPrompt?: string;
  /** Model to use (defaults to configured default model) */
  model?: string;
  /** Maximum tokens for completion */
  maxTokens?: number;
  /** Temperature (0-2) */
  temperature?: number;
  /** Top-p sampling */
  top_p?: number;
  /** Frequency penalty */
  frequency_penalty?: number;
  /** Presence penalty */
  presence_penalty?: number;
  /** Response format */
  responseFormat?: ChatCompletionResponseFormat;
  /** User identifier */
  user?: string;
  /** Conversation history */
  messages?: ChatCompletionMessageParam[];
  /** Whether to include ARCANOS routing message (default: true) */
  includeRoutingMessage?: boolean;
}

/**
 * Vision request parameters
 */
export interface VisionParams {
  /** User prompt describing the image */
  prompt: string;
  /** Base64-encoded image data */
  imageBase64: string;
  /** MIME type for image (default: image/png) */
  mimeType?: string;
  /** Model to use (defaults to vision model) */
  model?: string;
  /** Maximum tokens for completion */
  maxTokens?: number;
  /** Temperature (0-2) */
  temperature?: number;
  /** Image detail level */
  detail?: 'low' | 'high' | 'auto';
}

/**
 * Transcription request parameters
 */
export interface TranscriptionParams {
  /** Audio file data */
  audioFile: File | Blob | Buffer;
  /** Filename for the audio file */
  filename: string;
  /** Model to use (defaults to transcription model) */
  model?: string;
  /** Language code (optional) */
  language?: string;
  /** Response format */
  responseFormat?: 'json' | 'text' | 'srt' | 'verbose_json' | 'vtt';
  /** Temperature */
  temperature?: number;
}

/**
 * Image generation request parameters
 */
export interface ImageParams {
  /** Text prompt describing the desired image */
  prompt: string;
  /** Image size */
  size?: ImageSize;
  /** Model to use (defaults to image generation model) */
  model?: string;
  /** Quality setting */
  quality?: 'standard' | 'hd';
  /** Number of images to generate */
  n?: number;
  /** Response format */
  responseFormat?: 'url' | 'b64_json';
}

/**
 * Embedding request parameters
 */
export interface EmbeddingParams {
  /** Text to embed */
  input: string | string[];
  /** Model to use */
  model: string;
  /** User identifier */
  user?: string;
}

interface LegacyUsageShape {
  input_tokens?: number;
  output_tokens?: number;
  total_tokens?: number;
}

function normalizeMessageContent(content: unknown): string {
  if (typeof content === 'string') {
    return content;
  }

  if (Array.isArray(content)) {
    return content
      .map((part) => {
        if (!part || typeof part !== 'object') {
          return '';
        }
        const typedPart = part as Record<string, unknown>;
        if (typedPart.type === 'text' && typeof typedPart.text === 'string') {
          return typedPart.text;
        }
        if (typedPart.type === 'input_text' && typeof typedPart.text === 'string') {
          return typedPart.text;
        }
        return '';
      })
      .filter((part) => part.length > 0)
      .join('\n');
  }

  return '';
}

function extractUsage(usage: unknown): { promptTokens: number; completionTokens: number; totalTokens: number } {
  const typedUsage = (usage ?? {}) as LegacyUsageShape;
  const promptTokens = Number.isFinite(typedUsage.input_tokens) ? Number(typedUsage.input_tokens) : 0;
  const completionTokens = Number.isFinite(typedUsage.output_tokens) ? Number(typedUsage.output_tokens) : 0;
  const totalTokens = Number.isFinite(typedUsage.total_tokens)
    ? Number(typedUsage.total_tokens)
    : promptTokens + completionTokens;
  return {
    promptTokens,
    completionTokens,
    totalTokens
  };
}

/**
 * Build a Responses API payload from chat-style params.
 *
 * Purpose:
 * - Canonicalize runtime request construction on Responses API.
 * Inputs/Outputs:
 * - Input: chat-oriented params used across legacy call sites.
 * - Output: OpenAI Responses non-streaming payload.
 * Edge cases:
 * - Empty/unsupported message parts are normalized to text.
 */
export function buildResponsesRequest(
  params: ChatParams
): ResponseCreateParamsNonStreaming {
  const {
    prompt,
    systemPrompt,
    model,
    maxTokens = ROUTING_MAX_TOKENS,
    temperature = OPENAI_COMPLETION_DEFAULTS.TEMPERATURE,
    top_p = OPENAI_COMPLETION_DEFAULTS.TOP_P,
    messages,
    includeRoutingMessage = true,
    responseFormat,
    user
  } = params;

  let preparedMessages: ChatCompletionMessageParam[] =
    messages && messages.length > 0 ? [...messages] : buildSystemPromptMessages(prompt, systemPrompt);

  //audit Assumption: routing guard message must be prepended exactly once; risk: missing routing behavior or duplicate instructions; invariant: one routing system message present when enabled; handling: detect then prepend if absent.
  if (includeRoutingMessage) {
    const hasRoutingMessage = preparedMessages.some(
      (message) =>
        message.role === 'system' &&
        typeof message.content === 'string' &&
        message.content.includes(ARCANOS_ROUTING_MESSAGE)
    );
    if (!hasRoutingMessage) {
      preparedMessages = [{ role: 'system', content: ARCANOS_ROUTING_MESSAGE }, ...preparedMessages];
    }
  }

  const instructionText = preparedMessages
    .filter((message) => message.role === 'system')
    .map((message) => normalizeMessageContent(message.content))
    .filter((value) => value.length > 0)
    .join('\n\n');

  const responseInput = preparedMessages
    .filter((message) => message.role !== 'system')
    .map((message) => {
      const contentText = normalizeMessageContent(message.content);
      return {
        role: message.role === 'assistant' ? 'assistant' : 'user',
        content: [{ type: 'input_text', text: contentText.length > 0 ? contentText : prompt }]
      };
    });

  const tokenParameters = getTokenParameter(model || 'gpt-4.1-mini', maxTokens);
  const maxOutputTokens = tokenParameters.max_completion_tokens || tokenParameters.max_tokens || maxTokens;

  const payload: ResponseCreateParamsNonStreaming = {
    model: model || 'gpt-4.1-mini',
    input: (responseInput.length > 0
      ? responseInput
      : [{ role: 'user', content: [{ type: 'input_text', text: prompt }] }]) as unknown as ResponseInput,
    temperature,
    top_p,
    max_output_tokens: maxOutputTokens
  };

  if (instructionText.length > 0) {
    payload.instructions = instructionText;
  }

  //audit Assumption: legacy callers may still request structured JSON; risk: behavior drift when ignored; invariant: format hints forwarded when recognizable; handling: map supported response_format types to Responses text.format.
  if (responseFormat && typeof responseFormat === 'object' && 'type' in responseFormat) {
    const responseType = String((responseFormat as { type?: unknown }).type || '').toLowerCase();
    if (responseType === 'json_object') {
      payload.text = { format: { type: 'json_object' } };
    } else if (responseType === 'json_schema') {
      const jsonSchema = (responseFormat as { json_schema?: unknown }).json_schema;
      payload.text = {
        format: {
          type: 'json_schema',
          ...(jsonSchema && typeof jsonSchema === 'object' ? { json_schema: jsonSchema } : {})
        } as never
      };
    }
  }

  if (user) {
    payload.metadata = { user };
  }

  return payload;
}

/**
 * Build a Responses API payload for vision analysis.
 *
 * @param params - Vision request parameters.
 * @returns Responses API payload.
 */
export function buildVisionResponsesRequest(
  params: VisionParams
): ResponseCreateParamsNonStreaming {
  const {
    prompt,
    imageBase64,
    mimeType = 'image/png',
    model = 'gpt-4o',
    maxTokens = ROUTING_MAX_TOKENS,
    temperature = OPENAI_COMPLETION_DEFAULTS.TEMPERATURE
  } = params;

  return {
    model,
    input: [
      {
        role: 'user',
        content: [
          { type: 'input_text', text: prompt },
          {
            type: 'input_image',
            image_url: `data:${mimeType};base64,${imageBase64}`
          }
        ]
      }
    ] as unknown as ResponseInput,
    temperature,
    max_output_tokens: maxTokens
  };
}

/**
 * Extract text content from a Responses API response.
 *
 * @param response - Responses API response payload.
 * @param fallback - Fallback text when no output text is present.
 * @returns Normalized output text.
 */
export function extractResponseOutputText(response: OpenAIResponse, fallback = ''): string {
  const typedOutputText = (response as { output_text?: unknown }).output_text;
  if (typeof typedOutputText === 'string' && typedOutputText.trim().length > 0) {
    return typedOutputText.trim();
  }

  const outputItems = Array.isArray(response.output) ? response.output : [];
  for (const outputItem of outputItems) {
    if (!outputItem || typeof outputItem !== 'object') {
      continue;
    }
    const typedOutputItem = outputItem as unknown as Record<string, unknown>;
    const contentItems = Array.isArray(typedOutputItem.content) ? typedOutputItem.content : [];
    for (const contentItem of contentItems) {
      if (!contentItem || typeof contentItem !== 'object') {
        continue;
      }
      const typedContentItem = contentItem as Record<string, unknown>;
      if (typedContentItem.type === 'output_text' && typeof typedContentItem.text === 'string') {
        const normalizedText = typedContentItem.text.trim();
        if (normalizedText.length > 0) {
          return normalizedText;
        }
      }
    }
  }

  return fallback;
}

/**
 * Convert a Responses API response into a legacy ChatCompletion shape.
 *
 * Purpose:
 * - Preserve compatibility with existing consumers while migrating internals.
 * Inputs/Outputs:
 * - Input: Responses API response + requested model.
 * - Output: ChatCompletion-compatible object.
 * Edge cases:
 * - Missing usage metadata falls back to zero-valued token counts.
 */
export function convertResponseToLegacyChatCompletion(
  response: OpenAIResponse,
  requestedModel: string
): OpenAI.Chat.Completions.ChatCompletion {
  const outputText = extractResponseOutputText(response, '');
  const usage = extractUsage(response.usage);
  const createdSource = (response as { created_at?: unknown }).created_at;
  const created = typeof createdSource === 'number'
    ? Math.floor(createdSource)
    : Math.floor(Date.now() / 1000);

  return {
    id: response.id || `legacy_${Date.now()}`,
    object: 'chat.completion',
    created,
    model: response.model || requestedModel,
    choices: [
      {
        index: 0,
        message: {
          role: 'assistant',
          content: outputText,
          refusal: null
        },
        finish_reason: 'stop',
        logprobs: null
      }
    ],
    usage: {
      prompt_tokens: usage.promptTokens,
      completion_tokens: usage.completionTokens,
      total_tokens: usage.totalTokens
    }
  };
}

/**
 * Builds a chat completion request with ARCANOS routing message
 * 
 * Automatically prepends ARCANOS routing message to ensure proper
 * model routing and behavior. This is the standard way to create
 * chat completion requests in the codebase.
 * 
 * @param params - Chat completion parameters
 * @returns OpenAI chat completion request parameters
 */
export function buildChatCompletionRequest(
  params: ChatParams
): OpenAI.Chat.Completions.ChatCompletionCreateParamsNonStreaming {
  const {
    prompt,
    systemPrompt,
    model,
    maxTokens = ROUTING_MAX_TOKENS,
    temperature = OPENAI_COMPLETION_DEFAULTS.TEMPERATURE,
    top_p = OPENAI_COMPLETION_DEFAULTS.TOP_P,
    frequency_penalty = OPENAI_COMPLETION_DEFAULTS.FREQUENCY_PENALTY,
    presence_penalty = OPENAI_COMPLETION_DEFAULTS.PRESENCE_PENALTY,
    responseFormat,
    user,
    messages,
    includeRoutingMessage = true
  } = params;

  // Build messages array
  let preparedMessages: ChatCompletionMessageParam[];

  if (messages && messages.length > 0) {
    preparedMessages = [...messages];
  } else {
    preparedMessages = buildSystemPromptMessages(prompt, systemPrompt);
  }

  // Prepend ARCANOS routing message if requested
  if (includeRoutingMessage) {
    const hasRoutingMessage = preparedMessages.some(
      msg => msg.role === 'system' && 
      typeof msg.content === 'string' && 
      msg.content.includes(ARCANOS_ROUTING_MESSAGE)
    );

    if (!hasRoutingMessage) {
      preparedMessages = [
        { role: 'system', content: ARCANOS_ROUTING_MESSAGE },
        ...preparedMessages
      ];
    }
  }

  // Get token parameters for the model
  const tokenParams = getTokenParameter(model || 'gpt-4o-mini', maxTokens);

  // Build request payload
  const requestPayload: OpenAI.Chat.Completions.ChatCompletionCreateParamsNonStreaming = {
    model: model || 'gpt-4o-mini',
    messages: preparedMessages,
    stream: false,
    temperature,
    top_p,
    frequency_penalty,
    presence_penalty,
    ...tokenParams
  };

  if (responseFormat) {
    requestPayload.response_format = responseFormat;
  }

  if (user) {
    requestPayload.user = user;
  }

  return requestPayload;
}

/**
 * Builds a vision request for image analysis
 * 
 * Creates a properly formatted vision request with image data
 * and user prompt.
 * 
 * @param params - Vision request parameters
 * @returns OpenAI chat completion request parameters for vision
 */
export function buildVisionRequest(
  params: VisionParams
): OpenAI.Chat.Completions.ChatCompletionCreateParamsNonStreaming {
  const {
    prompt,
    imageBase64,
    mimeType = 'image/png',
    model = 'gpt-4o',
    maxTokens = ROUTING_MAX_TOKENS,
    temperature = OPENAI_COMPLETION_DEFAULTS.TEMPERATURE,
    detail = 'auto'
  } = params;

  const messages: ChatCompletionMessageParam[] = [
    {
      role: 'user',
      content: [
        { type: 'text', text: prompt },
        {
          type: 'image_url',
          image_url: {
            url: `data:${mimeType};base64,${imageBase64}`,
            detail
          }
        }
      ]
    }
  ];

  const tokenParams = getTokenParameter(model, maxTokens);

  return {
    model,
    messages,
    stream: false,
    temperature,
    ...tokenParams
  };
}

/**
 * Builds a transcription request for audio processing
 * 
 * Creates a properly formatted transcription request with audio file.
 * 
 * @param params - Transcription request parameters
 * @returns OpenAI transcription request parameters
 */
export function buildTranscriptionRequest(
  params: TranscriptionParams
): OpenAI.Audio.Transcriptions.TranscriptionCreateParamsNonStreaming {
  const {
    audioFile,
    filename,
    model = 'whisper-1',
    language,
    responseFormat = 'json',
    temperature
  } = params;

  const requestParams: OpenAI.Audio.Transcriptions.TranscriptionCreateParamsNonStreaming = {
    file: audioFile as File,
    model,
    response_format: responseFormat,
    stream: false
  };

  if (language) {
    requestParams.language = language;
  }

  if (temperature !== undefined) {
    requestParams.temperature = temperature;
  }

  return requestParams;
}

/**
 * Builds an image generation request
 * 
 * Creates a properly formatted image generation request.
 * 
 * @param params - Image generation parameters
 * @returns OpenAI image generation request parameters
 */
export function buildImageRequest(
  params: ImageParams
): OpenAI.Images.ImageGenerateParamsNonStreaming {
  const {
    prompt,
    size = DEFAULT_IMAGE_SIZE,
    model = IMAGE_GENERATION_MODEL,
    quality = 'standard',
    n = 1,
    responseFormat = 'b64_json'
  } = params;

  return {
    model,
    prompt,
    size,
    quality,
    n,
    response_format: responseFormat,
    stream: false
  };
}

/**
 * Builds an embedding request
 * 
 * Creates a properly formatted embedding request.
 * 
 * @param params - Embedding parameters
 * @returns OpenAI embedding request parameters
 */
export function buildEmbeddingRequest(
  params: EmbeddingParams
): OpenAI.Embeddings.EmbeddingCreateParams {
  const { input, model, user } = params;

  const requestParams: OpenAI.Embeddings.EmbeddingCreateParams = {
    model,
    input
  };

  if (user) {
    requestParams.user = user;
  }

  return requestParams;
}

/**
 * Default export for convenience
 */
export default {
  buildChatCompletionRequest,
  buildResponsesRequest,
  buildVisionRequest,
  buildVisionResponsesRequest,
  buildTranscriptionRequest,
  buildImageRequest,
  buildEmbeddingRequest,
  extractResponseOutputText,
  convertResponseToLegacyChatCompletion
};
